File: go/0983.minimum-cost-for-tickets/solution.go
// Created by none at 2024/10/01 14:41
// leetgo: 1.4.9
// https://leetcode.cn/problems/minimum-cost-for-tickets/

package main

import (
	"bufio"
	"fmt"
	"os"

	. "github.com/j178/leetgo/testutils/go"
)

// @lc code=begin

func mincostTickets(days []int, costs []int) int {
	/*
		f[i] = min(f[i-duration[i]]+cost[i])
	*/
	i, j, k := 0, 0, 0
	n := len(days)

	f := make([]int, n+1)

	for r, x := range days {
		for x-days[i] >= 1 {
			i++
		}
		for x-days[j] >= 7 {
			j++
		}
		for x-days[k] >= 30 {
			k++
		}
		f[r+1] = min(f[i]+costs[0], f[j]+costs[1], f[k]+costs[2])
	}
	return f[n]

}

// @lc code=end

func main() {
	stdin := bufio.NewReader(os.Stdin)
	days := Deserialize[[]int](ReadLine(stdin))
	costs := Deserialize[[]int](ReadLine(stdin))
	ans := mincostTickets(days, costs)

	fmt.Println("\noutput:", Serialize(ans))
}

File: python/0983.minimum-cost-for-tickets/solution.py
# Created by none at 2024/12/31 14:04
# leetgo: 1.4.11
# https://leetcode.com/problems/minimum-cost-for-tickets/

from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from functools import cache, cmp_to_key, lru_cache, reduce
from heapq import heapify, heappop, heappush, heappushpop, heapreplace
from itertools import accumulate, chain, count, pairwise, zip_longest
from math import ceil, comb, floor, gcd, inf, isqrt, lcm, log2, perm, sqrt
from operator import xor
from pprint import pprint
from string import ascii_lowercase
from typing import *
from typing import List, Optional

from leetgo_py import *

# @lc code=begin


class Solution:
    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        n = len(days)
        ticket = [1, 7, 30]

        @cache
        def dfs(i: int) -> int:
            if i == n:
                return 0
            res = inf
            for j, cost in enumerate(costs):
                k = bisect_left(days, days[i] + ticket[j])
                res = min(res, cost + dfs(k))
            return res

        return dfs(0)


# @lc code=end

if __name__ == "__main__":
    days: List[int] = deserialize("List[int]", read_line())
    costs: List[int] = deserialize("List[int]", read_line())
    ans = Solution().mincostTickets(days, costs)
    print("\noutput:", serialize(ans, "integer"))
